dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/thread/BackPressuredBlockingQueue.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.thread
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.core.configuration.BackPressureMitigation
+import com.datadog.android.core.configuration.BackPressureStrategy
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+
+internal class BackPressuredBlockingQueue<E : Any>(
+    private val logger: InternalLogger,
+    private val backPressureStrategy: BackPressureStrategy
+) : LinkedBlockingQueue<E>(
+    backPressureStrategy.capacity
+) {
+    override fun offer(e: E): Boolean {
+        return addWithBackPressure(e) {
+            @Suppress("UnsafeThirdPartyFunctionCall") // can't have NPE here
+            super.offer(it)
+        }
+    }
+
+    override fun offer(e: E, timeout: Long, unit: TimeUnit?): Boolean {
+        @Suppress("UnsafeThirdPartyFunctionCall") // can't have NPE here
+        val accepted = super.offer(e, timeout, unit)
+        if (!accepted) {
+            return offer(e)
+        } else {
+            if (remainingCapacity() == 0) {
```

Comment:
what is the proposed usage for the `onThresholdReached` callback? it is to have some logs on the customer side?

Reply:
It's something that our customer asked for: being able to disable some of their custom tracking if they reach this threshold for instance.

Reply:
ok, such use-case makes sense indeed. I guess they are not going to enable it back when it comes again below the threshold (we don't have anyway callback for that, when it comes)?